src/Stratis.Features.FederatedPeg/InputConsolidation/InputConsolidator.cs

```diff
@@ -0,0 +1,356 @@
+
+                List<ConsolidationTransaction> consolidationTransactions = new List<ConsolidationTransaction>();
+
+                while (oneRound.Count == FederatedPegSettings.MaxInputs)
+                {
+                    // We found a set of 50 that is worth enough so no more consolidation needed.
+                    if (oneRound.Sum(x => x.Transaction.Amount) >= amount + this.settings.GetWithdrawalTransactionFee(FederatedPegSettings.MaxInputs))
+                        break;
+
+                    // build a transaction and add it to our list.
+                    Transaction transaction = this.BuildConsolidatingTransaction(oneRound);
+
+                    // Something went wrong building transaction - start over. We will want to build them all from scratch in case wallet has changed state.
+                    if (transaction == null)
+                    {
+                        this.logger.LogDebug("Failure building specific consolidating transaction.");
+                        return null;
+                    }
+
+                    consolidationTransactions.Add(new ConsolidationTransaction
+                    {
+                        PartialTransaction = transaction,
+                        Status = ConsolidationTransactionStatus.Partial
+                    });
+
+                    roundNumber++;
+                    oneRound = unspentOutputs
+                        .Skip(roundNumber * FederatedPegSettings.MaxInputs)
+                        .Take(FederatedPegSettings.MaxInputs).ToList();
+                }
+
+                // Loop exits when we get a set of 50 that had a high enough amount, or when we run out of UTXOs aka a round less than 50
+
+                return consolidationTransactions;
+            }
+        }
+
+        /// <summary>
+        /// Build a consolidating transaction.
+        /// </summary>
+        private Transaction BuildConsolidatingTransaction(List<UnspentOutputReference> selectedInputs)
+        {
+            try
+            {
+                string walletPassword = this.walletManager.Secret.WalletPassword;
+                bool sign = (walletPassword ?? "") != "";
+
+                var multiSigContext = new TransactionBuildContext(new List<Recipient>())
+                {
+                    MinConfirmations = WithdrawalTransactionBuilder.MinConfirmations,
+                    Shuffle = false,
+                    IgnoreVerify = true,
+                    WalletPassword = walletPassword,
+                    Sign = sign,
+                    TransactionFee = FederatedPegSettings.ConsolidationFee,
+                    SelectedInputs = selectedInputs.Select(u => u.ToOutPoint()).ToList(),
+                    AllowOtherInputs = false,
+                    IsConsolidatingTransaction = true,
+                    Time = this.network.Consensus.IsProofOfStake 
+                        ? (uint?)selectedInputs.Max(x => x.Transaction.CreationTime).ToUnixTimeSeconds()
```

Comment:
Fix for https://stratisplatformuk.visualstudio.com/StratisBitcoinFullNode/_workitems/edit/3946